server/src/main/java/org/opensearch/wlm/cancellation/QueryGroupTaskCancellationService.java

```diff
@@ -0,0 +1,205 @@
+import org.opensearch.wlm.tracker.QueryGroupResourceUsageTrackerService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.opensearch.wlm.tracker.QueryGroupResourceUsageTrackerService.TRACKED_RESOURCES;
+
+/**
+ * Manages the cancellation of tasks enforced by QueryGroup thresholds on resource usage criteria.
+ * This class utilizes a strategy pattern through {@link MaximumResourceTaskSelectionStrategy} to identify tasks that exceed
+ * predefined resource usage limits and are therefore eligible for cancellation.
+ *
+ * <p>The cancellation process is initiated by evaluating the resource usage of each QueryGroup against its
+ * resource limits. Tasks that contribute to exceeding these limits are selected for cancellation based on the
+ * implemented task selection strategy.</p>
+ *
+ * <p>Instances of this class are configured with a map linking QueryGroup IDs to their corresponding resource usage
+ * views, a set of active QueryGroups, and a task selection strategy. These components collectively facilitate the
+ * identification and cancellation of tasks that threaten to breach QueryGroup resource limits.</p>
+ *
+ * @see MaximumResourceTaskSelectionStrategy
+ * @see QueryGroup
+ * @see ResourceType
+ */
+public class QueryGroupTaskCancellationService {
+    public static final double MIN_VALUE = 1e-9;
+
+    private final WorkloadManagementSettings workloadManagementSettings;
+    private final TaskSelectionStrategy taskSelectionStrategy;
+    private final QueryGroupResourceUsageTrackerService resourceUsageTrackerService;
+    // a map of QueryGroupId to its corresponding QueryGroupLevelResourceUsageView object
+    Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews;
+    private final Collection<QueryGroup> activeQueryGroups;
+    private final Collection<QueryGroup> deletedQueryGroups;
+
+    public QueryGroupTaskCancellationService(
+        WorkloadManagementSettings workloadManagementSettings,
+        TaskSelectionStrategy taskSelectionStrategy,
+        QueryGroupResourceUsageTrackerService resourceUsageTrackerService,
+        Collection<QueryGroup> activeQueryGroups,
+        Collection<QueryGroup> deletedQueryGroups
+    ) {
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.taskSelectionStrategy = taskSelectionStrategy;
+        this.resourceUsageTrackerService = resourceUsageTrackerService;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+    }
+
+    /**
+     * Cancel tasks based on the implemented strategy.
+     */
+    public final void cancelTasks(BooleanSupplier isNodeInDuress) {
+        queryGroupLevelResourceUsageViews = resourceUsageTrackerService.constructQueryGroupLevelUsageViews();
```

Comment:
How is this synchronized across concurrent invocations of `cancelTasks` method?

Reply:
this will not be called concurrently only periodically once in a second hence no synchronization is needed.

Reply:
Ideally it should be function parameter, since we are explicitly setting it. I would like to revisit this once we start using `QueryGroupTaskCancellationService`